python/ray/tune/progress_reporter.py

```diff
@@ -1532,3 +1534,67 @@ def _detect_progress_metrics(
         return None
 
     return getattr(trainable, "_progress_metrics", None)
+
+
+def _prepare_progress_reporter_for_ray_client(
+    progress_reporter: ProgressReporter,
+    verbosity: Union[int, Verbosity],
+    string_queue: Optional[Queue] = None,
+) -> Tuple[ProgressReporter, Queue]:
+    """Prepares progress reported for Ray Client by setting the string queue.
+
+    The string queue will be created if it's None."""
+    set_verbosity(verbosity)
+    progress_reporter = progress_reporter or _detect_reporter()
+
+    # JupyterNotebooks don't work with remote tune runs out of the box
+    # (e.g. via Ray client) as they don't have access to the main
+    # process stdout. So we introduce a queue here that accepts
+    # strings, which will then be displayed on the driver side.
+    if isinstance(progress_reporter, RemoteReporterMixin):
+        if string_queue is None:
+            string_queue = Queue(
+                actor_options={"num_cpus": 0, **_force_on_current_node(None)}
+            )
+        progress_reporter.output_queue = string_queue
+
+    return progress_reporter, string_queue
+
+
+def _get_remote_with_string_queue(
+    remote_future: ray.ObjectRef,
+    progress_reporter: ProgressReporter,
+    string_queue: Queue,
+) -> Any:
+    """Get result of remote future while checking the string queue."""
+    if string_queue is not None:
+
+        def get_next_queue_item():
+            try:
+                return string_queue.get(block=False)
+            except Empty:
+                return None
+
+    else:
+        # If we don't need a queue, use this dummy get fn instead of
+        # scheduling an unneeded actor
+        def get_next_queue_item():
+            return None
+
+    def _handle_string_queue():
+        string_item = get_next_queue_item()
+        while string_item is not None:
+            # This happens on the driver side
+            progress_reporter.display(string_item)
+
+            string_item = get_next_queue_item()
+
+    # ray.wait(...)[1] returns futures that are not ready, yet
+    while ray.wait([remote_future], timeout=0.2)[1]:
+        # Check if we have items to execute
+        _handle_string_queue()
+
+    # Handle queue one last time
+    _handle_string_queue()
+
+    return ray.get(remote_future)
```

Comment:
Good call moving this into a separate function! In this case we don't need the dummy queue handler anymore and can just return.

```suggestion
def _stream_client_output(
    remote_future: ray.ObjectRef,
    progress_reporter: ProgressReporter,
    string_queue: Queue,
) -> Any:
    """Stream items from string queue to progress_reporter until remote_future resolves""
    if string_queue is None:
        return

    def get_next_queue_item():
        try:
            return string_queue.get(block=False)
        except Empty:
            return None

    def _handle_string_queue():
        string_item = get_next_queue_item()
        while string_item is not None:
            # This happens on the driver side
            progress_reporter.display(string_item)

            string_item = get_next_queue_item()

    # ray.wait(...)[1] returns futures that are not ready, yet
    while ray.wait([remote_future], timeout=0.2)[1]:
        # Check if we have items to execute
        _handle_string_queue()

    # Handle queue one last time
    _handle_string_queue()
```